File: smalivm/src/main/java/org/cf/smalivm/opcode/InvokeOp.java
package org.cf.smalivm.opcode;

import gnu.trove.list.TIntList;
import gnu.trove.list.linked.TIntLinkedList;
import org.cf.smalivm.*;
import org.cf.smalivm.context.*;
import org.cf.smalivm.dex.CommonTypes;
import org.cf.smalivm.emulate.MethodEmulator;
import org.cf.smalivm.type.*;
import org.cf.util.ClassNameUtils;
import org.cf.util.Utils;
import org.jf.dexlib2.builder.MethodLocation;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import javax.annotation.Nullable;
import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import java.util.List;
import java.util.Set;

public class InvokeOp extends ExecutionContextOp {

    private static final Logger log = LoggerFactory.getLogger(InvokeOp.class.getSimpleName());

    private final VirtualMethod method;
    private final int[] parameterRegisters;
    private final String[] analyzedParameterTypes;
    private final VirtualMachine vm;
    private final ClassManager classManager;
    private SideEffect.Level sideEffectLevel;

    InvokeOp(MethodLocation location, MethodLocation child, VirtualMethod method, int[] parameterRegisters,
             VirtualMachine vm) {
        super(location, child);
        this.method = method;
        this.parameterRegisters = parameterRegisters;
        analyzedParameterTypes = new String[method.getParameterTypeNames().size()];
        this.vm = vm;
        classManager = vm.getClassManager();
        sideEffectLevel = SideEffect.Level.STRONG;
    }

    @Override
    public void execute(ExecutionNode node, ExecutionContext context) {
        // TODO: In order to get working call stacks, refactor this to delegate most of the work to MethodExecutor.
        // This will remove InvokeOp as a weirdly complex op, and probably allow some methods to be made protected.
        // It also keeps things clear with method execution delegated to the class with the same name.
        // MethodExecutor can maintain a mapping such that calleeContext -> (callerContext, caller address)
        // With this mapping, stack traces can be reconstructed.

        MethodState callerMethodState = context.getMethodState();
        if (method.getSignature().equals(CommonTypes.OBJECT + "-><init>()V")) {
            // Object.<init> is a special little snow flake
            try {
                executeLocalObjectInit(callerMethodState);
            } catch (ClassNotFoundException | InstantiationException | IllegalAccessException | IllegalArgumentException | InvocationTargetException e) {
                log.error("Unexpected real exception initializing Object", e);
            }
            return;
        } else {
            if (method.getName().equals("clone") && parameterRegisters.length == 1) {
                int targetRegister = parameterRegisters[0];
                HeapItem item = context.getMethodState().peekRegister(targetRegister);
                String signature = method.getSignature();
                if (signature.charAt(0) == '[' || (!item.isNull() && item.getValueType().charAt(0) == '[')) {
                    // [Object;->clone()Ljava/lang/Object; is also a special snow flake
                    executeArrayClone(callerMethodState, node);
                    return;
                }
            }
        }

        analyzeParameterTypes(callerMethodState);

        // Have to do this at run time because robust type flow analysis is harder than just examining at run time
        VirtualMethod targetMethod = method;
        if (getName().startsWith("invoke-virtual") && !method.isFinal()) {
            int targetRegister = parameterRegisters[0];
            HeapItem item = context.getMethodState().peekRegister(targetRegister);
            targetMethod = resolveTargetMethod(item.getValue());
        }
        // Shouldn't reference method member now. Should use targetMethod for everything.
        String targetSignature = targetMethod.getSignature();

        // Try to reflect or emulate before executing local method.
        if (vm.getConfiguration().isSafe(targetSignature) || MethodEmulator.canEmulate(targetSignature)) {
            ExecutionContext calleeContext = buildNonLocalCalleeContext(context);
            boolean allArgumentsKnown = allArgumentsKnown(calleeContext.getMethodState());
            if (allArgumentsKnown || MethodEmulator.canHandleUnknownValues(targetSignature)) {
                executeNonLocalMethod(targetSignature, callerMethodState, calleeContext, node);
                return;
            } else {
                if (log.isTraceEnabled()) {
                    log.trace("Not emulating / reflecting {}; not all arguments are known", targetSignature);
                }
                assumeMaximumUnknown(callerMethodState);
                return;
            }
        }

        if (classManager.isFrameworkClass(targetMethod.getDefiningClass()) &&
                !classManager.isSafeFrameworkClass(targetMethod.getDefiningClass())) {
            if (log.isDebugEnabled()) {
                log.debug("Not executing unsafe framework method: {}. Assuming maximum ambiguity.", targetSignature);
            }
            assumeMaximumUnknown(callerMethodState);
            return;
        }

        if (!targetMethod.hasImplementation()) {
            if (log.isWarnEnabled()) {
                if (!targetMethod.isNative()) {
                    // This can happen if a method returns an object which implements an interface
                    // but the object is unknown, so the real virtual of the invocation target
                    // can't be determined. That's why this is a warning and not an error.
                    log.warn("Attempting to execute local native method without implementation: {}. Assuming maximum " +
                            "ambiguity.", targetSignature);
                } else {
                    log.warn("Cannot execute local native method: {}. Assuming maximum ambiguity.", targetSignature);
                }
            }
            assumeMaximumUnknown(callerMethodState);
            return;
        }

        ExecutionContext calleeContext = buildLocalCalleeContext(context, targetMethod);
        executeLocalMethod(targetSignature, context, calleeContext, node);
    }

    public int[] getParameterRegisters() {
        return parameterRegisters;
    }

    public String getReturnType() {
        return method.getReturnType();
    }

    @Override
    public SideEffect.Level getSideEffectLevel() {
        return sideEffectLevel;
    }

    public VirtualMethod getMethod() {
        return method;
    }

    public String[] getAnalyzedParameterTypes() {
        return analyzedParameterTypes;
    }

    @Override
    public String toString() {
        StringBuilder sb = new StringBuilder(getName());
        sb.append(" {");
        if (getName().contains("/range")) {
            sb.append("r").append(parameterRegisters[0]).append(" .. r")
                    .append(parameterRegisters[parameterRegisters.length - 1]);
        } else {
            if (parameterRegisters.length > 0) {
                for (int register : parameterRegisters) {
                    sb.append('r').append(register).append(", ");
                }
                sb.setLength(sb.length() - 2);
            }
        }
        sb.append("}, ").append(method);

        return sb.toString();
    }

    private void executeArrayClone(MethodState callerMethodState, ExecutionNode node) {
        int instanceRegister = parameterRegisters[0];
        HeapItem arrayItem = callerMethodState.peekRegister(instanceRegister);
        if (arrayItem.isUnknown()) {
            callerMethodState.assignResultRegister(new UnknownValue(), arrayItem.getType());
        } else if (arrayItem.isNull()) {
            // This operation would have thrown a null pointer exception, and nothing else.
            Throwable exception = new NullPointerException();
            addException(exception);
            node.clearChildren();
        } else {
            Method m = null;
            try {
                m = Object.class.getDeclaredMethod("clone");
                m.setAccessible(true);
                Object clone = m.invoke(arrayItem.getValue());
                callerMethodState.assignResultRegister(clone, arrayItem.getType());
            } catch (Exception e) {
                // TODO: should handle exceptions here and bubble them up
                e.printStackTrace();
            }
        }
    }

    private boolean allArgumentsKnown(MethodState mState) {
        for (int parameterRegister = mState.getParameterStart(); parameterRegister < mState.getRegisterCount(); ) {
            HeapItem item = mState.peekParameter(parameterRegister);
            if (item.isUnknown()) {
                return false;
            }
            String type = item.getType();
            parameterRegister += Utils.getRegisterSize(type);
        }

        return true;
    }

    private void analyzeParameterTypes(MethodState callerState) {
        /*
         * Type can be confused here. For example, creating a short, int, boolean, or *null* all appear:
         * const/4 v0,0x0 (could be true, (int)0, or (short)0, null, etc.)
         * If a more restrictive class is given in the method signature, prefer that, for example:
         * method argument is an int but signature declares it as boolean, so switch it to boolean
         * However, if the class is less specific, such as a super class or interface, do not use the less specific
         * class. For example:
         * method argument is Lchild_class; but signature says Lparent_class;, prefer Lchild_class;
         */
        List<String> parameterTypes = method.getParameterTypeNames();
        for (int i = 0; i < parameterRegisters.length; i++) {
            int callerRegister = parameterRegisters[i];
            HeapItem item = callerState.readRegister(callerRegister);
            String parameterTypeName = parameterTypes.get(i);
            String baseTypeName = item.getComponentBase();
            String type;
            if (item.isPrimitive() || ClassNameUtils.isPrimitive(baseTypeName)) {
                type = parameterTypeName;
            } else {
                VirtualType baseType = vm.getClassManager().getVirtualType(baseTypeName);
                VirtualType parameterType = vm.getClassManager().getVirtualType(parameterTypeName);
                if (baseType.getAncestors().contains(parameterType)) {
                    type = item.getType();
                } else {
                    type = parameterTypeName;
                }
            }
            analyzedParameterTypes[i] = type;
        }
    }

    private void assignCalleeMethodArguments(MethodState callerState, MethodState calleeState) {
        int parameterRegister = calleeState.getParameterStart();
        for (int i = 0; i < parameterRegisters.length; i++) {
            int callerRegister = parameterRegisters[i];
            HeapItem item = callerState.readRegister(callerRegister);
            String parameterType = analyzedParameterTypes[i];
            Object value = item.getValue();
            if (item.isPrimitive() && !item.isUnknown()) {
                boolean hasNullByteValue =
                        item.getType().equals("I") && value instanceof Number && item.asInteger() == 0;
                if (hasNullByteValue && ClassNameUtils.isObject(parameterType)) {
                    value = null;
                } else {
                    // The "I" type may actually be "S", "B", "C", etc. Cast to the given parameter type.
                    value = Utils.castToPrimitive(value, parameterType);
                }
            }
            HeapItem parameterItem = new HeapItem(value, parameterType);
            calleeState.assignParameter(parameterRegister, parameterItem);
            parameterRegister += Utils.getRegisterSize(parameterType);
        }
    }

    private void assumeMaximumUnknown(MethodState callerMethodState) {
        // TODO: add option to mark all class states unknown instead of just method state
        // This is heavy handed in most cases, but it's impossible to tell if the method we're
        // failing to execute modifies class state. If it did, we aren't capturing it here.
        boolean isInitializing = method.getSignature().contains(";-><init>(");
        for (int i = 0; i < method.getParameterTypeNames().size(); i++) {
            int register = parameterRegisters[i];
            HeapItem item = callerMethodState.readRegister(register);
            Object value = item.getValue();
            if (null == value) {
                // Null references can't mutate.
                continue;
            }

            String type;
            if (item.isUnknown()) {
                type = analyzedParameterTypes[i];
                if (!type.equals(item.getType())) {
                    VirtualType parameterType = vm.getClassManager().getVirtualType(type);
                    VirtualType argumentType = vm.getClassManager().getVirtualType(item.getType());

                    if (parameterType.isAncestorOf(argumentType)) {
                        type = item.getType();
                    }
                }
            } else {
                // If argument is known, use that type rather than relying on method signature.
                // Parameter type might be "Ljava/lang/Object;" but actual type is "Ljava/lang/String";
                type = item.getType();
            }

            if (!isInitializing) {
                // Even if immutable type, internal state can change in the initializer.
                if (vm.getConfiguration().isImmutable(type)) {
                    if (log.isTraceEnabled()) {
                        log.trace("{} is immutable", type);
                    }
                    continue;
                }
            }

            item = HeapItem.newUnknown(type);
            if (log.isDebugEnabled()) {
                log.debug("{} is mutable and passed into unresolvable method execution, making Unknown", type);
            }

            callerMethodState.pokeRegister(register, item);
        }

        if (isInitializing) {
            // TODO: If we're refusing to execute an <init> method, should create a new instance of at least the stub class or something
            // and update identities. That way we don't have weird Uninitialized instances floating around. Look at TestExceptionHandling
            // and how ExceptionalCode throws CustomException but Throwable isn't whitelisted.
        }

        if (!method.returnsVoid()) {
            HeapItem item = HeapItem.newUnknown(method.getReturnType());
            callerMethodState.assignResultRegister(item);
        }
    }

    private ExecutionContext buildLocalCalleeContext(ExecutionContext callerContext, VirtualMethod method) {
        ExecutionContext calleeContext = vm.spawnRootContext(method, callerContext, getAddress());
        MethodState callerMethodState = callerContext.getMethodState();
        MethodState calleeMethodState = calleeContext.getMethodState();
        assignCalleeMethodArguments(callerMethodState, calleeMethodState);

        // VirtualClass state merging is handled by the VM.

        return calleeContext;
    }

    private ExecutionContext buildNonLocalCalleeContext(ExecutionContext callerContext) {
        ExecutionContext calleeContext = new ExecutionContext(vm, method);
        int parameterSize = method.getParameterSize();
        int registerCount = parameterSize;
        MethodState calleeMethodState =
                new MethodState(calleeContext, registerCount, method.getParameterTypeNames().size(), parameterSize);
        assignCalleeMethodArguments(callerContext.getMethodState(), calleeMethodState);
        calleeContext.setMethodState(calleeMethodState);
        calleeContext.registerCaller(callerContext, getAddress());

        return calleeContext;
    }

    private void executeLocalMethod(String methodSignature, ExecutionContext callerContext,
                                    ExecutionContext calleeContext, ExecutionNode node) {
        ExecutionGraph graph = null;
        try {
            graph = vm.execute(methodSignature, calleeContext, callerContext, parameterRegisters);
        } catch (VirtualMachineException e) {
            log.warn(e.toString());
            if (e instanceof UnhandledVirtualException) {
                // An exception was thrown and there was no exception handler to catch it. It's not clear at this point
                // if the fault is in smalivm code or the app code.
                // TODO: bubble this up to the calling method
            }
        }

        if (graph == null) {
            // Maybe node visits or call depth exceeded?
            log.info("Problem executing {}, propagating ambiguity.", methodSignature);
            assumeMaximumUnknown(callerContext.getMethodState());

            return;
        }

        boolean hasOneNonThrow = false;
        for (int endAddress : graph.getConnectedTerminatingAddresses()) {
            Op endOp = graph.getTemplateNode(endAddress).getOp();
            if (endOp instanceof ThrowOp) {
                // At least one execution path leads to throwing an exception.
                Set<HeapItem> items = graph.getRegisterItems(endAddress, MethodState.ThrowRegister);
                for (HeapItem item : items) {
                    if (item.getValue() instanceof Throwable) {
                        Throwable exception = (Throwable) item.getValue();
                        addException(exception);
                    } else {
                        // Possibly UninitializedInstance
                        if (log.isWarnEnabled()) {
                            if (item.isUnknown()) {
                                log.warn("Method had possible execution path which throws an exception but cannot instantiate it because the value is unknown. Exception item: {}", item);
                            } else {
                                // May just need to whitelist Exception class
                                log.warn("Refusing to instantiate potentially unsafe thrown exception: {}.", item);
                            }
                        }
                    }
                }
            } else {
                hasOneNonThrow = true;
            }
        }

        if (!hasOneNonThrow) {
            // Everything was an exception; don't go execute any normal children
            node.clearChildren();
        } else {
            if (!method.getReturnType().equals(CommonTypes.VOID)) {
                // Terminating addresses may include throw ops which may not have a return register set
                TIntList addresses = new TIntLinkedList();
                for (int address : graph.getConnectedTerminatingAddresses()) {
                    if (graph.getOp(address) instanceof ReturnOp) {
                        addresses.add(address);
                    }
                }
                HeapItem consensus = graph.getRegisterConsensus(addresses.toArray(), MethodState.ReturnRegister);
                callerContext.getMethodState().assignResultRegister(consensus);
            } else {
                if (methodSignature.contains(";-><init>(")) {
                    // This was a call to a local parent <init> method
                    int calleeInstanceRegister = calleeContext.getMethodState().getParameterStart();
                    HeapItem newInstance = graph.getTerminatingRegisterConsensus(calleeInstanceRegister);
                    int instanceRegister = parameterRegisters[0];
                    callerContext.getMethodState().assignRegisterAndUpdateIdentities(instanceRegister, newInstance);
                }
            }
        }
        sideEffectLevel = graph.getHighestSideEffectLevel();
    }

    private void executeLocalObjectInit(MethodState callerMethodState) throws ClassNotFoundException, InstantiationException, IllegalAccessException, IllegalArgumentException, InvocationTargetException {
        int instanceRegister = parameterRegisters[0];
        HeapItem instanceItem = callerMethodState.peekRegister(instanceRegister);
        UninitializedInstance uninitializedInstance = (UninitializedInstance) instanceItem.getValue();
        VirtualType instanceType = uninitializedInstance.getType();

        // Create a Java class of the true type
        Class<?> klazz = vm.getClassLoader().loadClass(instanceType.getBinaryName());
        Object newInstance = ObjectInstantiator.newInstance(klazz);
        HeapItem newInstanceItem = new HeapItem(newInstance, instanceType.getName());
        callerMethodState.assignRegisterAndUpdateIdentities(instanceRegister, newInstanceItem);
    }

    private void executeNonLocalMethod(String methodDescriptor, MethodState callerMethodState,
                                       ExecutionContext calleeContext, ExecutionNode node) {
        if (MethodEmulator.canEmulate(methodDescriptor)) {
            MethodEmulator emulator = new MethodEmulator(vm, calleeContext, methodDescriptor);
            emulator.emulate(this);
            sideEffectLevel = emulator.getSideEffectLevel();
            if (emulator.getExceptions().size() > 0) {
                node.clearChildren();
                node.setExceptions(emulator.getExceptions());
                return;
            }
        } else if (vm.getConfiguration().isSafe(methodDescriptor)) {
            MethodReflector reflector = new MethodReflector(vm, method);
            try {
                reflector.reflect(calleeContext.getMethodState()); // playa play
            } catch (Exception e) {
                node.setException(e);
                node.clearChildren();
                return;
            }

            // Only safe, non-side-effect methods are allowed to be reflected.
            sideEffectLevel = SideEffect.Level.NONE;
        }

        if (!method.isStatic()) {
            // This is virtual and the instance parse may have been initialized or mutated.
            HeapItem originalInstanceItem = callerMethodState.peekRegister(parameterRegisters[0]);
            HeapItem newInstanceItem = calleeContext.getMethodState().peekParameter(0);
            if (originalInstanceItem.getValue() != newInstanceItem.getValue()) {
                // Instance has been initialized, i.e. was UninitializedInstance
                // Use assignRegisterAndUpdateIdentities because multiple registers may have an identical
                // UninitializedInstance, and those need to be updated with the new instance.
                callerMethodState.assignRegisterAndUpdateIdentities(parameterRegisters[0], newInstanceItem);
            } else {
                boolean isMutable = !vm.getConfiguration().isImmutable(newInstanceItem.getType());
                if (isMutable) {
                    // The instance virtual is mutable so could have changed. Record that it was changed for the
                    // optimizer.
                    callerMethodState.assignRegister(parameterRegisters[0], newInstanceItem);
                }
            }
        }

        if (!method.returnsVoid()) {
            HeapItem returnItem = calleeContext.getMethodState().readReturnRegister();
            callerMethodState.assignResultRegister(returnItem);
        }
    }

    private
    @Nullable
    VirtualMethod resolveTargetMethod(Object virtualReference) {
        /*
         * A method may not be defined in the class referenced by invoke op. The method implementation may be part
         * of the super class. This method searches ancestor hierarchy for the class which implements the method.
         */
        VirtualType referenceType;
        if (virtualReference == null || virtualReference instanceof UnknownValue) {
            return method;
        }

        if (virtualReference instanceof UninitializedInstance) {
            UninitializedInstance instance = (UninitializedInstance) virtualReference;
            referenceType = instance.getType();
        } else {
            String targetType = ClassNameUtils.toInternal(virtualReference.getClass());
            referenceType = classManager.getVirtualType(targetType);
        }

        VirtualMethod targetMethod = referenceType.getMethod(method.getDescriptor());
        if (targetMethod != null && targetMethod.hasImplementation()) {
            return targetMethod;
        }

        return method;
    }

}

File: smalivm/src/main/java/org/cf/smalivm/opcode/InvokeOp.kt
package org.cf.smalivm.opcode

import org.cf.smalivm.SideEffectLevel
import org.cf.smalivm.configuration.Configuration
import org.cf.smalivm.dex.CommonTypes
import org.cf.smalivm.dex.SmaliClassLoader
import org.cf.smalivm.type.ClassManager
import org.cf.smalivm.type.UninitializedInstance
import org.cf.smalivm.type.UnknownValue
import org.cf.smalivm.type.VirtualMethod
import org.cf.smalivm2.*
import org.cf.util.ClassNameUtils
import org.cf.util.Utils
import org.jf.dexlib2.builder.MethodLocation
import org.jf.dexlib2.iface.instruction.Instruction
import org.jf.dexlib2.iface.instruction.ReferenceInstruction
import org.jf.dexlib2.iface.instruction.formats.Instruction35c
import org.jf.dexlib2.iface.instruction.formats.Instruction3rc
import org.jf.dexlib2.iface.reference.MethodReference
import org.jf.dexlib2.util.ReferenceUtil
import org.slf4j.LoggerFactory
import java.util.*

class InvokeOp internal constructor(
    location: MethodLocation,
    val method: VirtualMethod,
    val parameterRegisters: IntArray,
) : Op(location) {

//    var isDebugMode: Boolean
//    var debuggedMethodExecutor: MethodExecutor? = null
//        private set
//    private var debuggedCallerContext: ExecutionContext? = null
//    private var debuggedCalleeContext: ExecutionContext? = null
//    private var debuggedNode: ExecutionNode? = null

//    init {
//        analyzedParameterTypes = arrayOfNulls(method.parameterTypeNames.size)
//        // TODO: set these in the node!
////        sideEffectLevel = SideEffect.Level.STRONG
//        isDebugMode = false
//    }

    override val registersReadCount = parameterRegisters.size
    override val registersAssignedCount = 1

    override fun execute(node: ExecutionNode): Array<out UnresolvedChild> {
//        if (method.signature == CommonTypes.OBJECT + "-><init>()V") {
//            // Object.<init> is a special little snow flake
//            return executeLocalObjectInit(node)
//        }
//
//        if (method.name == "clone" && parameterRegisters.size == 1 && method.signature[0] == '[') {
//            val targetRegister = parameterRegisters[0]
//            val value = node.state.peekRegister(targetRegister)!!
//            if (value.isNotNull && value.rawType[0] == '[') {
//                // Ljava/lang/Object;->clone()Ljava/lang/Object; is also a special snow flake
//                return executeObjectArrayClone(node)
//            }
//        }

        // Must do this at run time to take advantage of extra type information
        val targetMethod = if (!method.isFinal && (name.startsWith("invoke-virtual") || name.startsWith("invoke-interface"))) {
            val targetRegister = parameterRegisters[0]
            val target = node.state.peekRegister(targetRegister)!!
            resolveTargetMethod(target.raw, node.classManager)
        } else {
            method
        }

        val analyzedParameterTypes = analyzeParameterTypes(node)
        val calleeState = ExecutionState.build(targetMethod, node.classManager, node.classLoader, node.configuration)
        assignCalleeMethodArguments(node.state, calleeState, analyzedParameterTypes)
        return callMethod(targetMethod, calleeState)
    }

    val returnType: String
        get() = method.returnType

    override fun toString(): String {
        val sb = StringBuilder(name)
        sb.append(" {")
        if (name.contains("/range")) {
            sb.append('r').append(parameterRegisters[0])
                .append(" .. r").append(parameterRegisters[parameterRegisters.size - 1])
        } else {
            if (parameterRegisters.isNotEmpty()) {
                for (register in parameterRegisters) {
                    sb.append('r').append(register).append(", ")
                }
                sb.setLength(sb.length - 2)
            }
        }
        sb.append("}, ").append(method)
        return sb.toString()
    }

//    private fun executeObjectArrayClone(node: ExecutionNode): Array<out UnresolvedChild> {
//        val instanceRegister = parameterRegisters[0]
//        val arrayItem = node.state.peekRegister(instanceRegister)!!
//        if (arrayItem.isNull) {
//            // This operation would have thrown a null pointer exception, and nothing else.
//            return throwException(NullPointerException::class.java)
//        }
//        val clone = when {
//            arrayItem.isUnknown -> UnknownValue()
//            else -> {
//                try {
//                    val m = Any::class.java.getDeclaredMethod("clone")
//                    m.isAccessible = true
//                    m.invoke(arrayItem.raw)
//                } catch (e: Exception) {
//                    log.error("Real exception initializing Object; returning unknown", e)
//                    UnknownValue()
//                }
//            }
//        }
//        node.state.assignResultRegister(clone, arrayItem.type)
//        return finishOp()
//    }

    private fun analyzeParameterTypes(node: ExecutionNode): Array<String> {
        /*
         * Type can be confused here. For example, creating a short, int, boolean, or *null* all appear as:
         * const/4 v0, 0x0
         * Use type in method signature if it's more restrictive. For example, if the method argument is
         * an int but the method signature declares it as a boolean, use boolean.
         * However, if the class is less specific, such as a super class or interface, do not use the less specific
         * class. For example:
         * method argument is Lchild_class; but signature says Lparent_class;, prefer Lchild_class;
         */
        val parameterTypes = method.parameterTypeNames
        val analyzedParameterTypes = Array(parameterTypes.size) { parameterTypes[it] }
        for (i in parameterRegisters.indices) {
            val argument = node.state.readRegister(parameterRegisters[i])
            val parameterTypeName = parameterTypes[i]
            val baseTypeName = argument.componentBase
            val type = if (parameterTypeName == CommonTypes.OBJECT && argument.isArray) {
                argument.type
            } else if (argument.isPrimitive || ClassNameUtils.isPrimitive(baseTypeName)) {
                parameterTypeName
            } else {
                val baseType = node.classManager.getVirtualType(baseTypeName)
                val parameterType = node.classManager.getVirtualType(parameterTypeName)
                when {
                    baseType.ancestors.contains(parameterType) -> argument.type
                    else -> parameterTypeName
                }
            }
            analyzedParameterTypes[i] = type
        }

        return analyzedParameterTypes
    }

    private fun assignCalleeMethodArguments(callerState: ExecutionState, calleeState: ExecutionState, analyzedParameterTypes: Array<String>) {
        var parameterRegister = calleeState.firstParameterRegister
        for (i in parameterRegisters.indices) {
            val callerRegister = parameterRegisters[i]
            val value = callerState.readRegister(callerRegister)
            val parameterType = analyzedParameterTypes[i]
            var raw = value.raw
            if (value.isPrimitive && !value.isUnknown) {
                val hasNullByteValue = value.type == CommonTypes.INTEGER && raw is Number && value.toInteger() == 0
                raw = if (hasNullByteValue && ClassNameUtils.isObject(parameterType)) {
                    null
                } else {
                    // The "I" type may actually be "S", "B", "C", "Z", etc. Cast to the given parameter type.
                    Utils.castToPrimitive(raw, parameterType)
                }
            }
            val argument = Value.wrap(raw, parameterType)
            calleeState.assignParameter(parameterRegister, argument)
            parameterRegister += Utils.getRegisterSize(parameterType)
        }
    }

    override fun resume(node: ExecutionNode, calleeGraph: ExecutionGraph2): Array<out UnresolvedChild> {
        if (calleeGraph.aborted) {
            // VM either aborted execution because of some limit or it couldn't be reflected or emulated.
            assumeMaximumUnknown(node, calleeGraph.method)
            return finishOp()
        }

        // Mutable parameters may have been changed. Need to merge changes back into caller.
        // Reflection mutates parameters directly. No need to merge them ourselves.
        if (!calleeGraph.reflected) {
            val parameterTypes = calleeGraph.method.parameterTypeNames
            var parameterRegister = calleeGraph.root.state.firstParameterRegister
            for (parameterIndex in parameterTypes.indices) {
                val type = parameterTypes[parameterIndex]
                if (node.configuration.isMutable(type)) {
                    val value = calleeGraph.getTerminatingParameterConsensus(parameterRegister)
                    val register = parameterRegisters[parameterIndex]
                    node.state.assignParameter(register, value, updateIdentities = true)
                }
                parameterRegister += Utils.getRegisterSize(type)
            }
        }

        val hasOneReturnPath = calleeGraph.hasAtLeastOneReturnPath()
        if (hasOneReturnPath) {
            if (!method.returnsVoid()) {
                // Terminating addresses may include throw ops which may not have a return register set
                val result = calleeGraph.getRegisterConsensus(calleeGraph.connectedReturnAddresses, ExecutionState.RETURN_REGISTER)
                node.state.assignResultRegister(result)
            } else {
                if (calleeGraph.method.name == "<init>") {
                    // This was a call to a local parent <init> method
                    val calleeInstanceRegister = calleeGraph.root.state.firstParameterRegister
                    val newInstance = calleeGraph.getRegisterConsensus(calleeGraph.terminatingAddresses, calleeInstanceRegister)
                    val instanceRegister = parameterRegisters[0]
                    node.state.assignRegister(instanceRegister, newInstance, updateIdentities = true)
                }
            }
        }
        node.sideEffectLevel = calleeGraph.getHighestSideEffectLevel()

        // Unhandled exceptions in callee are handled by VM when the step is finished
        return if (hasOneReturnPath) {
            finishOp()
        } else {
            // This method never returns so it must only ever throw exceptions.
            finishOpWithoutChildren()
        }
    }

    private fun assumeMaximumUnknown(node: ExecutionNode, calledMethod: VirtualMethod) {
        // TODO: add option to mark all class states unknown instead of just method state
        // This is heavy handed in most cases, and maybe there's a way to optimize. It's hard to tell if the method we're
        // failing to execute modifies class state. If it did, we aren't capturing it here.
        node.sideEffectLevel = SideEffectLevel.STRONG
        for (i in parameterRegisters.indices) {
            val register = parameterRegisters[i]
            val arg = node.state.readRegister(register)
            if (arg.isNull) {
                // Null is immutable. No need to mark unknown.
                continue
            }
            if (i == 0 && calledMethod.name == "<init>") {
                // Instance initialization failed so mark unknown. Note this happens even if arg is immutable because state can change in <init>
                node.state.pokeRegister(register, Value.unknown(arg.type))
                continue
            } else if (arg.isImmutable) {
                // Could not have possibly mutated in aborted method.
                continue
            }
            node.state.pokeRegister(register, Value.unknown(arg.type))
        }

        if (!calledMethod.returnsVoid()) {
            val result = Value.unknown(calledMethod.returnType)
            node.state.assignResultRegister(result)
        }

        // TODO: add tests for this
        if (calledMethod.name == "<clinit>") {
            node.state.setClassInitialized(calledMethod.definingClass, SideEffectLevel.STRONG)
            for (field in calledMethod.definingClass.fields) {
                val value = Value.unknown(field.type)
                node.state.pokeField(field, value)
            }
        }
    }

    private fun resolveTargetMethod(virtualReference: Any?, classManager: ClassManager): VirtualMethod {
        /*
         * A method may not be defined in the class referenced by invoke op. The method implementation may be part
         * of the super class. This method searches ancestor hierarchy for the class which implements the method.
         */
        if (virtualReference == null || virtualReference is UnknownValue) {
            return method
        }
        val referenceType = if (virtualReference is UninitializedInstance) {
            virtualReference.type
        } else {
            val targetType = ClassNameUtils.toInternal(virtualReference.javaClass)
            classManager.getVirtualType(targetType)
        }
        val targetMethod = referenceType.getMethod(method.descriptor)

        return if (targetMethod != null && targetMethod.hasImplementation()) {
            targetMethod
        } else {
            method
        }
    }

    companion object : OpFactory {
        private val log = LoggerFactory.getLogger(InvokeOp::class.java.simpleName)
        private fun buildParameterRegisters(parameterTypes: List<String>, registers: IntArray): IntArray {
            val parameterRegisters = LinkedList<Int>()
            var index = 0
            for (parameterType in parameterTypes) {
                parameterRegisters.add(registers[index])
                index += Utils.getRegisterSize(parameterType)
            }
            return parameterRegisters.toIntArray()
        }

        private fun buildRegisters(instr: Instruction?): IntArray {
            return when (instr) {
                is Instruction3rc -> buildRegisters3rc(instr)
                else -> buildRegisters35c(instr as Instruction35c)
            }
        }

        private fun buildRegisters35c(instruction: Instruction35c?): IntArray {
            val registerCount = instruction!!.registerCount
            val registers = IntArray(registerCount)
            when (registerCount) {
                5 -> {
                    registers[4] = instruction.registerG
                    registers[3] = instruction.registerF
                    registers[2] = instruction.registerE
                    registers[1] = instruction.registerD
                    registers[0] = instruction.registerC
                }
                4 -> {
                    registers[3] = instruction.registerF
                    registers[2] = instruction.registerE
                    registers[1] = instruction.registerD
                    registers[0] = instruction.registerC
                }
                3 -> {
                    registers[2] = instruction.registerE
                    registers[1] = instruction.registerD
                    registers[0] = instruction.registerC
                }
                2 -> {
                    registers[1] = instruction.registerD
                    registers[0] = instruction.registerC
                }
                1 -> registers[0] = instruction.registerC
            }
            return registers
        }

        private fun buildRegisters3rc(instruction: Instruction3rc): IntArray {
            val registerCount = instruction.registerCount
            val start = instruction.startRegister
            val end = start + registerCount
            val registers = IntArray(registerCount)
            for (i in start until end) {
                registers[i - start] = i
            }
            return registers
        }

        override fun build(
            location: MethodLocation,
            classManager: ClassManager,
            classLoader: SmaliClassLoader,
            configuration: Configuration
        ): Op {
            val instruction = location.instruction as ReferenceInstruction
            val methodReference = instruction.reference as MethodReference
            val registers = buildRegisters(location.instruction)
            val className = methodReference.definingClass
            val type = classManager.getVirtualType(className)
            val methodSignature = ReferenceUtil.getMethodDescriptor(methodReference)
            val methodDescriptor = methodSignature.split("->")[1]
            val method = type.getMethod(methodDescriptor) ?: throw RuntimeException("Method doesn't exist: $methodSignature")
            val parameterRegisters = buildParameterRegisters(method.parameterTypeNames, registers)
            return InvokeOp(location, method, parameterRegisters)
        }
    }
}
